classes/local/util/manager.php

```diff
@@ -154,6 +154,40 @@ protected function prepare_files_skeletons() {
                 $this->prepare_file_skeleton('db/upgradelib.php', 'php_internal_file', 'db_upgradelib');
             }
         }
+
+        if ($this->should_have('observers')) {
+            $this->prepare_file_skeleton('db/events.php', 'php_internal_file', 'db_events');
+            $this->prepare_observers();
+        }
+    }
+
+    /**
+     * Prepares the observer class files.
+     */
+    protected function prepare_observers() {
+
+        foreach ($this->recipe['observers'] as $observer) {
+            if (empty($observer['eventname'])) {
+                throw new exception('Missing eventname from observers');
+            }
+
+            if (empty($observer['callback'])) {
+                throw new exception('Missing callback from observers');
+            }
+
+            $observerrecipe = $this->recipe;
+            $observerrecipe['observer'] = $observer;
+
+            $observernamespace = explode('\\', $observer['callback']);
```

Comment:
I don't think it is always the case to have callbacks implemented as methods in an autoloaded class like this. What if the recipe file defines the callback be a regular function in some locallib.php file, such as `quiz_attempt_submitted_handler()` defined in mod/quiz/db/events.php

The implementation should be changed so that both backslash and double colon is optional in the callback name.

Also note, the plugin author can make it in a way that all callbacks are static methods of a single file. It is once again M:N situation. You can even have multiple events observed by a single callback, if I remember correctly.

So here we must prepare a list of files, and each of them can have a class with method(s) and/or functions.

Maybe there should be an abstract skel class for all PHP files that would have methods like `add_function()` and `add_class()` that would allow us to dynamically prepare contents of the files like `locallib.php` or `mod/quiz/classes/group_observers.php`